src/analysis/vector/geometry_checker/qgsgeometrymissingvertexcheck.cpp

```diff
@@ -0,0 +1,119 @@
+#include "qgscurvepolygon.h"
+#include "qgscurve.h"
+#include "qgslinestring.h"
+#include "qgsgeometryengine.h"
+#include "qgsgeometryutils.h"
+
+void QgsGeometryMissingVertexCheck::collectErrors( QList<QgsGeometryCheckError *> &errors, QStringList &messages, QAtomicInt *progressCounter, const QMap<QString, QgsFeatureIds> &ids ) const
+{
+  Q_UNUSED( messages );
+  if ( progressCounter )
+    progressCounter->fetchAndAddRelaxed( 1 );
+
+  QMap<QString, QgsFeatureIds> featureIds = ids.isEmpty() ? allLayerFeatureIds() : ids;
+
+  QgsFeaturePool *featurePool = mContext->featurePools.value( featureIds.firstKey() );
+
+  const QgsGeometryCheckerUtils::LayerFeatures layerFeatures( mContext->featurePools, featureIds, mCompatibleGeometryTypes, nullptr, mContext, true );
+
+  for ( const QgsGeometryCheckerUtils::LayerFeature &layerFeature : layerFeatures )
+  {
+    const QgsAbstractGeometry *geom = layerFeature.geometry().constGet();
+
+    if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( geom ) )
+    {
+      processPolygon( polygon, featurePool, errors, layerFeature );
+    }
+    else if ( QgsGeometryCollection *collection = qgsgeometry_cast<QgsGeometryCollection *>( geom ) )
+    {
+      const int numGeometries = collection->numGeometries();
+      for ( int i = 0; i < numGeometries; ++i )
+      {
+        if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( collection->geometryN( i ) ) )
+        {
+          processPolygon( polygon, featurePool, errors, layerFeature );
+        }
+      }
+    }
+  }
+}
+
+void QgsGeometryMissingVertexCheck::fixError( QgsGeometryCheckError *error, int method, const QMap<QString, int> &mergeAttributeIndices, Changes &changes ) const
+{
+  Q_UNUSED( mergeAttributeIndices );
+  Q_UNUSED( changes );
+  if ( method == NoChange )
+  {
+    error->setFixed( method );
+  }
+}
+
+QStringList QgsGeometryMissingVertexCheck::resolutionMethods() const
+{
+  static QStringList methods = QStringList() << tr( "No action" );
+  return methods;
+}
+
+void QgsGeometryMissingVertexCheck::processPolygon( const QgsCurvePolygon *polygon, QgsFeaturePool *featurePool, QList<QgsGeometryCheckError *> &errors, const QgsGeometryCheckerUtils::LayerFeature &layerFeature ) const
+{
+  std::unique_ptr<QgsGeometryCollection> boundaries = qgis::make_unique<QgsGeometryCollection>();
+
+  std::unique_ptr< QgsGeometryEngine > geomEngine = QgsGeometryCheckerUtils::createGeomEngine( polygon->exteriorRing(), mContext->tolerance );
+  boundaries->addGeometry( geomEngine->buffer( mContext->tolerance, 5 ) );
+
+  const int numRings = polygon->numInteriorRings();
+  for ( int i = 0; i < numRings; ++i )
+  {
+    geomEngine = QgsGeometryCheckerUtils::createGeomEngine( polygon->exteriorRing(), mContext->tolerance );
+    boundaries->addGeometry( geomEngine->buffer( mContext->tolerance, 5 ) );
+  }
+
+  geomEngine = QgsGeometryCheckerUtils::createGeomEngine( boundaries.get(), mContext->tolerance );
```

Comment:
You should prepare this engine -- it'll make the later call to intersects much faster